ambry-account/src/main/java/com/github/ambry/account/mysql/MySqlDataAccessor.java

```diff
@@ -182,24 +232,116 @@ void onSuccess(OperationType operationType, long operationTimeInMs) {
   }
 
   /**
-   * Reset to initial state.
+   * Close the active connection and clear the statement cache.
    * This should be called after a failed database operation.
    */
   synchronized void reset() {
     for (PreparedStatement statement : statementCache.values()) {
-      try {
-        statement.close();
-      } catch (SQLException e) {
-        logger.error("Closing prepared statement", e);
-      }
+      closeQuietly(statement);
     }
     statementCache.clear();
-    if (activeConnection != null) {
+    closeQuietly(activeConnection);
+  }
+
+  /**
+   * Connect to the best available database instance that matches the specified criteria.
+   * Order of preference for instances to connect to:
+   *   1) writeable instance in local colo
+   *   2) writable instance in any colo
+   *   3) readonly instance in local colo (if needWritable is false)
+   *   4) readonly instance in any colo (if needWritable is false)
+   * @param needWritable whether the endpoint needs to be writable
+   * @return a pair of {@link DbEndpoint} and corresponding {@link Connection}.
+   * @throws SQLException if connection could not be made to a suitable endpoint.
+   */
+  private Pair<DbEndpoint, Connection> connectToBestAvailableEndpoint(boolean needWritable) throws SQLException {
+    SQLException lastException = null;
+    for (DbEndpoint dbEndpoint : sortedDbEndpoints) {
+      if (!isBetterEndpoint(dbEndpoint, connectedEndpoint)) {
+        // This is the best we can do
+        if (needWritable && !connectedEndpoint.isWriteable()) {
+          // TODO: throw lastException if any
+          throw new SQLException("Could not connect to any writable database");
+        } else {
+          return new Pair<>(connectedEndpoint, activeConnection);
+        }
+      }
+      if (needWritable && !dbEndpoint.isWriteable()) {
+        continue;
```

Comment:
Would returning here be okay too? Since db end points are sorted by writable property, remaining ones might be read-only ones.

Reply:
Changed to throw exception